runtime/near-vm/engine/src/universal/code_memory.rs

```diff
@@ -2,170 +2,254 @@
-
-        // 2. Allocate the pages. Mark them all read-write.
-
-        self.mmap = Mmap::with_at_least(total_len)?;
-
-        // 3. Determine where the pointers to each function, executable section
-        // or data section are. Copy the functions. Collect the addresses of each and return them.
-
-        let mut bytes = 0;
-        let mut buf = self.mmap.as_mut_slice();
-        for func in functions {
-            let len = round_up(Self::function_allocation_size(*func), ARCH_FUNCTION_ALIGNMENT);
-            let (func_buf, next_buf) = buf.split_at_mut(len);
-            buf = next_buf;
-            bytes += len;
-
-            let vmfunc = Self::copy_function(*func, func_buf);
-            assert_eq!(vmfunc.as_ptr() as usize % ARCH_FUNCTION_ALIGNMENT, 0);
-            function_result.push(vmfunc);
-        }
-        for section in executable_sections {
-            let section = &section.bytes;
-            assert_eq!(buf.as_mut_ptr() as usize % ARCH_FUNCTION_ALIGNMENT, 0);
-            let len = round_up(section.len(), ARCH_FUNCTION_ALIGNMENT);
-            let (s, next_buf) = buf.split_at_mut(len);
-            buf = next_buf;
-            bytes += len;
-            s[..section.len()].copy_from_slice(*section);
-            executable_section_result.push(s);
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            // SAFETY: We have made sure that this is the only reference to the memory region by
+            // requiring a mutable self reference.
+            std::slice::from_raw_parts_mut(self.map, self.size)
         }
+    }
 
-        self.start_of_nonexecutable_pages = bytes;
-
-        if !data_sections.is_empty() {
-            // Data sections have different page permissions from the executable
-            // code that came before it, so they need to be on different pages.
-            let padding = round_up(bytes, page_size) - bytes;
-            buf = buf.split_at_mut(padding).1;
-
-            for section in data_sections {
-                let section = &section.bytes;
-                assert_eq!(buf.as_mut_ptr() as usize % DATA_SECTION_ALIGNMENT, 0);
-                let len = round_up(section.len(), DATA_SECTION_ALIGNMENT);
-                let (s, next_buf) = buf.split_at_mut(len);
-                buf = next_buf;
-                s[..section.len()].copy_from_slice(*section);
-                data_section_result.push(s);
+    /// Ensure this CodeMemory is at least of the requested size.
+    ///
+    /// This may invalidate any data previously written into the mapping.
+    pub fn resize(mut self, size: usize) -> rustix::io::Result<Self> {
+        if self.size < size {
+            let source_pool = self.source_pool.take();
+            // Ideally we would use mremap, but see
+            // https://bugzilla.kernel.org/show_bug.cgi?id=8691
+            unsafe {
+                mm::munmap(self.map.cast(), self.size)?;
+                std::mem::forget(self);
```

Comment:
Instead of trickery with `std::mem::forget`, could we instead just `drop` it? That would be equivalent, right?

I'm torn between the two strategies, not sure which I like better. Both are juggling with deconstructors.
I guess your code looks better with the comment regarding `mremap`. But it adds another `unsafe` section that we could avoid otherwise.

Any other reasons why to prefer one over the other?

Reply:
`drop` does not allow for error handling, which manual `munmap` and `forget` allows.

Reply:
oh, good point, yeah I missed the `?`